koap/build.gradle.kts

```diff
@@ -53,3 +53,15 @@ npmPublishing {
         }
     }
 }
+
+tasks.register("run-js-validation-testing") {
+
+    description = "Runs the package for testing"
+    group = "Verification"
+    dependsOn("assemble")
+
+    doLast {
+        var response = runCommand("../gradlew -Pnpm.publish.repository.github.authToken=fakingit -Pnpm.publish.version=0.0.3-test24 assemble && cd koap_js_tests && npm uninstall @juullabs/koap && npm install file://../build/publications/npm/js && npm run test")
```

Comment:
Does `npm install` not overwrite an existing install? Do you have to `uninstall` first every time?

Reply:
Honestly not sure. I'll try it out. @earthtosid @davidtaylor-juul Are there any reasons that we need to uninstall? I can't remember if that came up specifically on Friday.

Reply:
Is `assemble` packaging the npm as a single file?
If so, couldn't we run the test without having to deal the with the install?
https://stackoverflow.com/a/28739855

Reply:
Running without the uninstall works fine, but I'll leave that decision to more experienced minds.

As for the assemble question: which npm/script are you referring to? The js api output or the test code?

Reply:
The npm system has aggressive caching and will not overwrite if the version number has not changed, hence the advice to uninstall/reinstall. There are other ways to force it such as removing the package lock file that may or may not be faster if performance is a concern (it probably isn't).

Regarding the change here to use `npm install` I can explain. The previous iteration of these tests was using the Webpack artifacts (a single js file), but this was a hack due to the inability to load the non-webpack artifacts which are the ones that actually ship in the npm package (a structured npm multi-file/folder package). There is a very strong preference to test the module that actually ships. This is particularly important now that we are using npm shrink wrap hacks to access the bundled dependencies as it may affect visibility of those dependencies.

To do so, it is necessary to import the module that is emitted via the npm publish task. Due to security restrictions with the node runtime it is not possible to simply import the code from the build artifact folder (you cannot escape to the parent folder via `../`). So it then becomes necessary to perform the npm install using a `file://` url in order to be able to import the module into the javascript tests.

Reply:
> As for the assemble question: which npm/script are you referring to? The js api output or the test code?

David answered my question.

So, it sounds like we should drop a lot of the shell commands and just depend on the `packJsNpmPublication` Gradle task.

Then you could have something like:

```kotlin
tasks.register("publishToLocalNpm") {
    dependsOn("packJsNpmPublication")
    runCommand("npm install -force file://../build/publications/npm/js")
}

tasks.register("apiTest") {
    dependsOn("publishToLocalNpm")
    runCommand("npm run tests")
}
```

Reply:
Uhh. packJsNpmPublication--itself--fails because of "unspecified version."

`
Justins-MBP:koap justin.tiffner$ ./gradlew packJsNpmPublication

> Configure project :koap
Kotlin Multiplatform Projects are an Alpha feature. See: https://kotlinlang.org/docs/reference/evolution/components-stability.html. To hide this message, add 'kotlin.mpp.stability.nowarn=true' to the Gradle properties.

NPM Repository [github] is invalid. Skipping...

> Task :koap:packJsNpmPublication FAILED
npm ERR! Invalid version: "unspecified"

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/justin.tiffner/.npm/_logs/2020-11-18T00_46_26_101Z-debug.log
`